#### Clustring vs Classification

In [13]:
import joblib
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [14]:
# Load  split data with clusters

X_train_with_clusters = joblib.load('../data/processed/X_train_with_cluster.pkl')
X_test_with_clusters = joblib.load('../data/processed/X_test_with_cluster.pkl')


# Load  split data without clusters

X_train_no_clusters = joblib.load('../data/processed/X_train.pkl')
X_test_no_clusters = joblib.load('../data/processed/X_test.pkl')
y_train_no_clusters = joblib.load('../data/processed/y_train.pkl')
y_test_no_clusters = joblib.load('../data/processed/y_test.pkl')

In [15]:
# train model with clusters
model_clustering = XGBClassifier(
        objective="binary:logistic",
        eval_metric="logloss",
        random_state=42,
        n_estimators=200,
        learning_rate=0.01,
        colsample_bytree= 1.0,
        max_depth= 3,
        scale_pos_weight=6,
        subsample= 1.0
    )
model_clustering.fit(X_train_with_clusters, y_train_no_clusters)
y_pred_clustering = model_clustering.predict(X_test_with_clusters)
accuracy_clustering = accuracy_score(y_test_no_clusters, y_pred_clustering)
print("Accuracy with clusters:", accuracy_clustering)
print("Classification Report with clusters:\n", classification_report(y_test_no_clusters, y_pred_clustering))


Accuracy with clusters: 0.6306049822064057
Classification Report with clusters:
               precision    recall  f1-score   support

           0       0.96      0.52      0.67      1033
           1       0.41      0.94      0.57       372

    accuracy                           0.63      1405
   macro avg       0.69      0.73      0.62      1405
weighted avg       0.82      0.63      0.65      1405



In [16]:
# train model without clusters
model_clustering = XGBClassifier(
        objective="binary:logistic",
        eval_metric="logloss",
        random_state=42,
        n_estimators=200,
        learning_rate=0.01,
        colsample_bytree= 1.0,
        max_depth= 3,
        scale_pos_weight=6,
        subsample= 1.0
    )
model_clustering.fit(X_train_no_clusters, y_train_no_clusters)
y_pred_clustering = model_clustering.predict(X_test_no_clusters)
accuracy_clustering = accuracy_score(y_test_no_clusters, y_pred_clustering)
print("Accuracy without clusters:", accuracy_clustering)
print("Classification Report without clusters:\n", classification_report(y_test_no_clusters, y_pred_clustering))


Accuracy without clusters: 0.6405693950177936
Classification Report without clusters:
               precision    recall  f1-score   support

           0       0.97      0.53      0.68      1033
           1       0.42      0.95      0.58       372

    accuracy                           0.64      1405
   macro avg       0.69      0.74      0.63      1405
weighted avg       0.82      0.64      0.66      1405



#### rapport entre les deux modeles avec et sans (clustring)

**Conclusion** : L'ajout de l'information issue de la segmentation (clusters KMeans) n'a pas amélioré les performances de prédiction du modèle. Au contraire, le modèle initial a enregistré de meilleurs scores sur l'ensemble des métriques, notamment une baisse du Recall pour la classe cible (de 0.95 à 0.94) et de la précision. Les clusters ont probablement agi comme du bruit (noise) pour l'algorithme XGBoost. Par conséquent, l'information issue du clustering ne sera pas retenue pour le modèle final.